lib/chef/resource/macos_userdefaults.rb

```diff
@@ -102,149 +104,63 @@ class MacosUserDefaults < Chef::Resource
-
-          shell_out!(cmd, user: new_resource.user)
+          write_preferences(new_resource)
         end
       end
 
       action :delete, description: "Delete a key from a domain." do
         # if it's not there there's nothing to remove
         return unless current_resource
 
-        converge_by("delete domain:#{new_resource.domain} key:#{new_resource.key}") do
+        # TODO: implement in CF and use here
+      end
 
-          cmd = defaults_modify_cmd
-          Chef::Log.debug("Removing defaults key by shelling out: #{cmd.join(" ")}")
+      action_class do
+        CF_MAPPING = {
+          current_user: CF::Preferences::CURRENT_USER,
+          all_users: CF::Preferences::ALL_USERS,
+          current_host: CF::Preferences::CURRENT_HOST,
+          all_hosts: CF::Preferences::ALL_HOSTS,
+          current: CF::Preferences::CURRENT_HOST # TODO: deprecation warning for this option
+        }
+
+        def read_preferences(new_resource)
+          CF::Preferences.get(new_resource.key, new_resource.domain, mapped_user, mapped_host)
+        end
 
-          shell_out!(cmd, user: new_resource.user)
+        def write_preferences(new_resource)
+          CF::Preferences.set(new_resource.key, new_resource.value, new_resource.domain, mapped_user, mapped_host)
         end
-      end
 
-      action_class do
-        #
-        # The command used to write or delete delete values from domains
-        #
-        # @return [Array] Array representation of defaults command to run
-        #
-        def defaults_modify_cmd
-          cmd = ["/usr/bin/defaults"]
-
-          if new_resource.host == :current
-            cmd.concat(["-currentHost"])
-          elsif new_resource.host # they specified a non-nil value, which is a hostname
-            cmd.concat(["-host", new_resource.host])
-          end
-
-          cmd.concat([action.to_s, new_resource.domain, new_resource.key])
-          cmd.concat(processed_value) if action == :write
-          cmd.prepend("sudo") if new_resource.sudo
-          cmd
+        def mapped_user
+          CF_MAPPING[valid_user.to_sym] || valid_user.to_s
         end
 
-        #
-        # convert the provided value into the format defaults expects
-        #
-        # @return [array] array of values starting with the type if applicable
-        #
-        def processed_value
-          type = new_resource.type || value_type(new_resource.value)
-
-          # when dict this creates an array of values ["Key1", "Value1", "Key2", "Value2" ...]
-          cmd_values = ["-#{type}"]
-
-          case type
-          when "dict"
-            cmd_values.concat(new_resource.value.flatten)
-          when "array"
-            cmd_values.concat(new_resource.value)
-          when "bool"
-            cmd_values.concat(bool_to_defaults_bool(new_resource.value))
-          else
-            cmd_values.concat([new_resource.value])
-          end
-
-          cmd_values
```

Comment:
it is difficult to review this for correctness since this behavior has been shoved inside the gem.  we should probably have a customer that actually uses the existing code try all this before shipping it.